File: kind-of-utf8-validation/kouv.go
package main

import (
	"fmt"
)

func byteToDigit(b byte) []int {
	bits := make([]int, 8) // A byte has 8 bits

	for i := 7; i >= 0; i-- {
		mask := byte(1 << i)

		if (b & mask) != 0 {
			bits[7-i] = 1
		} else {
			bits[7-i] = 0
		}
	}
	return bits
}

func validation(bytes []byte) bool {
	if len(bytes) == 0 {
		return true
	}

	bbs := byteToDigit(bytes[0])

	// 10 at the beginning
	if bbs[0] == 1 && bbs[1] == 0 {
		return false
	}

	if bbs[0] == 0 && bbs[1] == 1 {
		return validation(bytes[1:])
	}

	if bbs[0] == 0 && bbs[1] == 0 {
		return false
	}

	byteLen := 0
	for _, b := range bbs {
		if b == 1 {
			byteLen += 1
		} else {
			break
		}
	}

	if byteLen > len(bytes) {
		return false
	}

	for _, x := range bytes[1:byteLen] {
		xx := byteToDigit(x)
		if xx[0] != 1 || xx[1] != 0 {
			return false
		}
	}

	return validation(bytes[byteLen:])

}

type testCase struct {
	bytes     []byte
	validated bool
}

func main() {
	fmt.Println(byteToDigit(0b10000000))

	testcases := []testCase{
		{
			bytes:     []byte{0b10000000},
			validated: false,
		},
		{
			bytes:     []byte{0b01000000, 0b10000000},
			validated: false,
		},
		{
			bytes:     []byte{0b01000000, 0b01000000},
			validated: true,
		},
		{
			bytes:     []byte{0b11000000},
			validated: false,
		},
		{
			bytes:     []byte{0b11000000, 0b10101111},
			validated: true,
		},
		{
			bytes:     []byte{0b01000000, 0b11100000, 0b10101111, 0b10101111},
			validated: true,
		},
		{
			bytes:     []byte{0b01000000, 0b11100000, 0b10101111, 0b10101111, 0b01111111},
			validated: true,
		},
	}

	for ind, testc := range testcases {
		fmt.Println("testcase: ", ind+1)
		fmt.Printf("should be %v, actually %v\n", testc.validated, validation(testc.bytes))

	}
}

File: kind-of-utf8-validation/kouv.rs
/// some special utf8
/// 01000000
/// 11000000 (means two bytes) 10xxxxxx
/// 11100000 (means three bytes) 10xxxxxx 10xxxxxx

#[derive(PartialEq, Eq, Debug)]
enum State {
    ZeroOne,
    OneZero,
    MoreBytes(u8),
    Error,
}

fn get_state(x: &u8) -> State {
    let a = (0b10000000 & *x) != 0;
    let b = (0b01000000 & *x) != 0;

    match (a, b) {
        (true, true) => {
            let mut all = 2;
            for i in (0..6).rev() {
                if (1 << i) & *x != 0 {
                    all += 1
                } else {
                    break;
                }
            }
            State::MoreBytes(all)
        }
        (true, false) => State::OneZero,
        (false, true) => State::ZeroOne,
        (false, false) => State::Error,
    }
}

fn validation(inputs: &[u8]) -> Result<bool, String> {
    if inputs.len() == 0 {
        return Ok(true);
    }

    inputs.iter().for_each(|b| {
        println!("Byte {:08b}", b);
    });
    println!("");

    let more_bytes = match get_state(&inputs[0]) {
        State::ZeroOne => return validation(&inputs[1..]),
        State::OneZero => return Ok(false),
        State::MoreBytes(n) => n,
        State::Error => return Err("error".to_string()),
    };

    if more_bytes as usize > inputs.len() {
        return Ok(false);
    }

    if inputs
        .get(1..(more_bytes as usize))
        .unwrap()
        .iter()
        .any(|x| {
            if let State::OneZero = get_state(x) {
                false
            } else {
                true
            }
        })
    {
        return Ok(false);
    }

    validation(&inputs[(more_bytes as usize)..])
}

fn main() {
    //println!("{:?}", 0b11111111)
    assert_eq!(get_state(&0b10000000), State::OneZero);
    assert_eq!(get_state(&0b10100000), State::OneZero);
    assert_eq!(get_state(&0b10110000), State::OneZero);
    assert_eq!(get_state(&0b01110000), State::ZeroOne);
    assert_eq!(get_state(&0b01000000), State::ZeroOne);
    assert_eq!(get_state(&0b11110000), State::MoreBytes(4));

    let testcases = vec![
        (false, vec![0b01000000, 0b10000000]),
        (false, vec![0b10000000]),
        (false, vec![0b11000000]),
        (true, vec![0b11000000, 0b10101111]),
        (true, vec![0b01000000, 0b11100000, 0b10101111, 0b10101111]),
        (
            true,
            vec![0b01000000, 0b11100000, 0b10101111, 0b10101111, 0b01111111],
        ),
    ];

    for (ind, (res, testcase)) in testcases.into_iter().enumerate() {
        println!("");
        println!("test case {}", ind + 1);
        assert_eq!(validation(&testcase), Ok(res));
    }
}
